features-set-computer/src/main/scala/FeaturesSetComputer.scala

```diff
@@ -0,0 +1,106 @@
+ * a given set of language constructs
+ *
+ * @param availableFeatures the features that can be included in the computed set
+ */
+class FeaturesSetComputer(availableFeatures: List[Feature]) {
+
+  /**
+   * Computes the minimal set of Features needed by a WhitelistChecker to allow all the nodes in the given tree
+   *
+   * @param tree the tree whose nodes should be allowed
+   * @return the said minimal set of Features, wrapped in a Some, if it exists, otherwise None
+   */
+  def minimalFeaturesSetFor(tree: Tree): Option[Set[Feature]] = {
+    minimalFeaturesSetFor(tree.collect { case tree => tree })
+  }
+
+  /**
+   * Assuming that a WhitelistChecker with the set of Features S1 rejected a program outputting the provided set of
+   * Violations, computes the minimal set S2 of Features s.t. a WhitelistChecker with with features S1 U S2
+   * will accept the program
+   *
+   * @param violations the violations reported by the checker with the incomplete set of features (S1)
+   * @return the set (S2) of additionally required features, wrapped in a Some, if it exists, otherwise None
+   */
+  def minimalFeaturesSetToResolve(violations: List[Violation]): Option[Set[Feature]] = {
+    minimalFeaturesSetFor(violations.map(_.forbiddenNode))
+  }
+
+  /**
+   * Computes the minimal set of Features needed by a WhitelistChecker to allow all the provided nodes
+   *
+   * @param nodes the nodes to be allowed
+   * @return the said minimal set of Features, wrapped in a Some, if it exists, otherwise None
+   */
+  def minimalFeaturesSetFor(nodes: List[Tree]): Option[Set[Feature]] = {
+
+    val availableFeaturesCnt = availableFeatures.size
+    /**
+     * Array s.t. authorizations(n)(f) == true iff feature at index f allows node at index n
+     */
+    val authorizations = nodes.map(_ => new Array[Boolean](availableFeaturesCnt))
+    for ((node, nIdx) <- nodes.zipWithIndex) {
+      for ((feature, fIdx) <- availableFeatures.zipWithIndex) {
+        authorizations(nIdx)(fIdx) = WhitelistChecker(feature).checkTree(node).isEmpty
+      }
+    }
+
+    def allows(featIdx: Int, nodeIdx: Int): Boolean = authorizations(nodeIdx)(featIdx)
+
+    /**
+     * @param usedFeatures indices of the features that have already been included
+     * @param remNodes     indices of the nodes that are not covered by the used features
+     */
+    case class SearchThread(usedFeatures: Set[Int], remNodes: Set[Int]) {
+      def featureAdded(featIdx: Int): SearchThread = {
+        SearchThread(usedFeatures + featIdx, remNodes.filter(!allows(featIdx, _)))
+      }
+
+      def coversAllNodes: Boolean = remNodes.isEmpty
+
+      override def toString: String = {
+        s"SearchThread(${usedFeatures.map(availableFeatures(_))}, $remNodes)"
+      }
+    }
+
+    @tailrec
+    def search(currentThreads: List[SearchThread], currentFeatureIdx: Int): Option[Set[Int]] = {
```

Comment:
Could add a comment to explain roughly how the algorithm works?

Reply:
Done (0da3a0cf13115b3472999820bb52ea68aca3ab29)